### DISCUSSION TITLE
Cannot Read Cookies From Backend When I Send Request in getServerSideProps or getStaticProps

### DISCUSSION THREAD

**Discussion by yzhocaoglu** [vercel/next.js#22135] [Help] [answered]
I am creating a nextjs application with apollo graphql backend. What I am trying to do is reading cookies from backend.
Firstly, my code works fine when do request when page is loaded, like "clicking a button". But when I call my function from getServerSideProps or getStaticProps it doesnt work. I can make a request but backend cant read my cookies

also using next/cookies module where I make my request works well and can read the cookies.

why I am asking that question in nextjs discussions instead of 'cookie-parser' is my resolvers work well with graphql explorer and this issue is started when i connected my frontend.

As a solution I can give my cookies as parameters (I dont know what people calling them) to query (or mutation) and read params in backend.

but that is not what I want as a solution. Am I doing something wrong? or its not possible.
node version : v10.19.0
npm version: 6.14.8
yarn version: 1.22.10

apollo-server-express version : 2.20.0
cookie-parser version: 1.4.5

here is the code:

### Backend

resolver:

```javascript
Query: {
        me: async (_, __, { req, res }) => {
            console.log(req) // seems like i dont have any cookies
            if (!req.accID) {
                return {
                    status: false,
                    errorList: undefined,
                    account: undefined
                }
            }
            var acc = await FindByID(req.accID)
            return {
                status: true,
                errorList: undefined,
                account: acc
            }
        }
},
```

middleware i am using 

```javascript
app.use(async (req: any, res, next) => {
    var accessToken = req.cookies["accessToken"]
    var refreshToken = req.cookies["refreshToken"]

    if (!accessToken && !refreshToken) {
        return next()
    }

    try {
        var data = verify(accessToken, process.env.ACCESS_TOKEN_SECRET!) as any
        req.accID = data.id
        return next()
    } catch { }

    var data;

    try {
        data = verify(refreshToken, process.env.REFRESH_TOKEN_SECRET!) as any
    } catch { }

    var acc = await FindByID(data.id)
    if (acc.count != data.count) {
        return next()
    }
    req.accID = data.id

    var tokens = createToken(acc)
    setTokenAsCookie(tokens.accessToken, tokens.refreshToken, res)

    next()
})
```

### Frontend

where I run my request
```javascript
export async function getServerSideProps(ctx) {
  console.log("cookie", cookies(ctx)) // works fine
  var res = await Auth()
  return {
    props: {
      response: res
    }
  }
}
```

my auth() function

```javascript
export async function Auth() {
    console.log("auth!")
    var authQuery = await client.query({
        query: gql'
            {
                me {
                    status
                    errorList
                    account {
                        email
                        password
                        username
                    }
                }
            }
        '
    })
    console.log(authQuery.data.me)
    return authQuery.data.me.status
}
```

also sorry for my bad english

**Comment by ka8725** [comment]
I have the same problem: cookies are not bypassed from the AWS lambda function to node.js. Tested on next.js v10.0 (but actually the problem was there for earlier versions as well).

Here how I debugged it:

1. Run a nex.js app with some page that has this `getServerSideProps` function:
```
export const getServerSideProps = async (ctx) => {
  console.log(ctx.req.headers) // no cookies on vercel servers, locally works fine!

  return {
    props: { }
  }
}
```
2. Deploy it with vercel CLI
3. Watch logs: `vercel logs --scope <MY SCOPE> -f <MY APP URL>`
4. Go to browser or just `curl` with cookies
Result: `ctx.req.headers` has no `cookie` property set! Based on the headers coming to there it's clear that they have been rewritten by the AWS Lambda. So my guess is that something is going on there. The question is what and why?

**Comment by yzhocaoglu** [comment]
I did some research about how to send cookies as a header. I updated my `Auth()` function to following:
```javascript
export async function Auth(cookies) {
    var authQuery = await client.query({
        query: gql`
            {
                me {
                    status
                    errorList
                    account {
                        email
                        password
                        username
                    }
                }
            }
        `,
        context: {
            headers: {
                cookie: JSON.stringify(cookies)
            }
        }
    })
    console.log(authQuery.data.me)
    return authQuery.data.me.status
}
```

Now I can see cookies in backend. But cookie-parser still cant parse it. When I find what I am doing wrong I will post it in here.
Also I tried using `cookie:{...cookies}` but doesnt seems like working too.

**Comment by yzhocaoglu** [comment, marked as answer]
Quick update:
using that works fine and cookie-parser works too.
```javascript
context: {
    headers: {
        Cookie: `cookie1=${val}`
    }
}
```
